### ISSUE TITLE
Signature created from tuple type allows any number of arguments & doesn't preserve parameter names

### ISSUE THREAD

**Issue by bterlson** [Microsoft/TypeScript#25833]
**TypeScript Version:**  3.1.0-dev.20180719

**Code**

```ts
type R = {
  a: (x: number) => void;
  b: (x: string) => void;
};

type O = {
  emit<P extends keyof R>(x: P, ...args: Params<R[P]>): void;
};

type Params<F> = F extends (... args: infer T) => any ? T : never;

declare var x: O;
x.emit('a', 1, 2, 3); // expect an error here...
```

**Expected behavior:**
I would expect an error for x.emit as I am passing more parameters than should be allowed. Also, I would expect to see the the parameter names I've sliced off from R's members but those appear to be lost.

**Actual behavior:**
Emit's 2nd argument doesn't get a parameter name and there is no error for passing too many parameters.

**Playground Link:** [here](http://www.typescriptlang.org/play/index.html#src=type%20R%20%3D%20%7B%0D%0A%20%20a%3A%20(x%3A%20number)%20%3D%3E%20void%3B%0D%0A%20%20b%3A%20(x%3A%20string)%20%3D%3E%20void%3B%0D%0A%7D%3B%0D%0A%0D%0Atype%20O%20%3D%20%7B%0D%0A%20%20emit%3CP%20extends%20keyof%20R%3E(x%3A%20P%2C%20...args%3A%20Params%3CR%5BP%5D%3E)%3A%20void%3B%0D%0A%7D%3B%0D%0A%0D%0Atype%20Params%3CF%3E%20%3D%20F%20extends%20(...%20args%3A%20infer%20T)%20%3D%3E%20any%20%3F%20T%20%3A%20never%3B%0D%0A%0D%0Adeclare%20var%20x%3A%20O%3B%0D%0Ax.emit('a'%2C%201%2C%20)%3B%0D%0A)

**Comment by bterlson** [comment]
On further inspection, this is possibly related/identical to #25804. Replacing `R[P]` with a function type has the expected behavior.

**Comment by ahejlsberg** [comment]
As I suspected this appears to be fixed by #25817.

**Comment by bterlson** [comment]
Confirmed fixed (and unrelated to #25804).
